### PR TITLE
[Ehancement] Host support

### PR DESCRIPTION
**Pull request by mike-north**
Route-recognizer currently only supports same-origin requests. @stefanpenner and I paired on this to add support for CORS api mocking in our tests

### TODO
- [x] Tests!

### CODE REVIEW

**Comment by truenorth** [comment]
@trek this is ready for you to take a look at, whenever you have time

File: pretender.js [hunk 1]
```diff
@@ -5,18 +5,86 @@ var RouteRecognizer = isNode ? require('route-recognizer')['default'] : window.R
...
+function Registry(host) {
+  this.verbs = {
     GET: new RouteRecognizer(),
     PUT: new RouteRecognizer(),
     POST: new RouteRecognizer(),
     DELETE: new RouteRecognizer(),
     PATCH: new RouteRecognizer(),
-    HEAD: new RouteRecognizer()
+    HEAD: new RouteRecognizer(),
+    OPTIONS: new RouteRecognizer()
```
Comment (truenorth) [inline pretender.js]: not sure if anyone is doing anything crazy with options, but imo it doesn't do any harm to add it

File: pretender.js [hunk 2]
```diff
@@ -121,20 +189,6 @@ function verbify(verb){
...
-
-  if(HTTP_REGEXP.test(url)) {
-    var parser = window.document.createElement('a');
-    parser.href = url;
-
-    message = "Pretender will not respond to requests for URLs. It is not possible to accurately simluate the browser's CSP. "+
-              "Remove the " + parser.protocol +"//"+ parser.hostname +" from " + url + " and try again";
-    throw new Error(message)
-  }
-}
```
Comment (truenorth) [inline pretender.js]: @trek This seems to be directly in conflict with what I'm trying to do with this PR, so I just removed the check. Please let me know if you feel there's some way we can have our cake and eat it too.

**Comment by truenorth** [comment]
@trek ping

**Comment by cdunn** [comment]
Just what I needed, thanks for adding this @truenorth

**Comment by trek** [comment]
One "small" change.

I suspect previously we've coupled the idea of "request capture" and "server" because there's only ever been the notion of a single server. 

If we're going to support multiple servers, I'd prefer to make it explicit rather than implicit based on hostname. One thing that this PR doesn't address is server configuration. 

e.g.

```
let p = Pretender.new(() => {
  // multiple handlers for various servers
});

p.prepareBody = function(){}
```
 
It's _possible_ but not guaranteed that you'd want `prepareBody` to apply to all hosts defined in `p`. Much more likely that each host has its own specific behavior.

You've brushed lightly against the idea with the `Host` and `Registry` concepts, but I'd prefer it fleshed out a bit more and made explicit:

```
import Host from "pretender/hosts";
import Pretender from "pretender";

adServer = new Host(() => {
  this.get(...);
});
adServer.prepareBody = function(){};

export default new Pretender(() => {
  this.host('http://ads.mymegacorp.org/', adServer);
});
```

Make sense?

**Comment by truenorth** [comment]
@trek sounds good, and I think that having a distinct `Pretender` per host makes more sense when thinking about re-using servers across many tests. I'll see if I can find some time this week to refactor this to align more with this vision

**Comment by samselikoff** [comment]
v. exciting! anything I can do to help?

**Comment by trek** [comment]
Looks like this needs some love (and a rebase!)

**Comment by lessless** [comment]
@truenorth can you please rebase?

**Comment by stefanpenner** [comment]
@truenorth ping ping

**Comment by jamesdixon** [comment]
@truenorth rebase, rebase! :+1:

**Comment by tomoguisuru** [comment]
@truenorth any chance of getting a rebase on this soon?

**Comment by truenorth** [comment]
Attempting to rebase, but #73 is clashing with this, causing a test failure. Trying to dig deeper to determine what's going on

**Comment by truenorth** [comment]
Rebase is complete. There is an intermittently failing test (described in #87), but I don't believe it has anything to do with this code.

**Comment by stefanpenner** [comment]
WOOHOOO!

**Comment by mike-north** [comment]
:+1:

**Comment by samselikoff** [comment]
:tada: :tada: :tada:

**Comment by jamesdixon** [comment]
Well done, gents!

**Pull request merged**